applications/GeoMechanicsApplication/custom_processes/geo_extrapolate_integration_point_values_to_nodes_process.h

```diff
@@ -0,0 +1,131 @@
+    std::map<const Variable<Vector>*, Vector>         mZeroValuesOfVectorVariables;
+    std::map<const Variable<Matrix>*, Matrix>         mZeroValuesOfMatrixVariables;
+
+    void FillVariableLists(const Parameters& rParameters);
+    void InitializeVectorAndMatrixZeros();
+    void InitializeZerosOfVectorVariables(Element&           rFirstElement,
+                                          SizeType           NumberOfIntegrationPoints,
+                                          const ProcessInfo& rProcessInfo);
+    void InitializeZerosOfMatrixVariables(Element&           rFirstElement,
+                                          SizeType           NumberOfIntegrationPoints,
+                                          const ProcessInfo& rProcessInfo);
+    void InitializeVariables();
+    void InitializeAverageVariablesForElements() const;
+
+    template <class T>
+    bool TryAddVariableToList(const std::string& rVariableName, std::vector<const Variable<T>*>& rList) const
+    {
+        const bool variable_is_of_correct_type = KratosComponents<Variable<T>>::Has(rVariableName);
+        if (variable_is_of_correct_type) {
+            auto& thisVariable = KratosComponents<Variable<T>>::Get(rVariableName);
+            rList.push_back(&thisVariable);
+        }
+
+        return variable_is_of_correct_type;
+    }
+
+    template <class T, class U>
+    void AddIntegrationContributionsToNodes(Element&           rElement,
+                                            const Variable<T>& rVariable,
+                                            const Matrix&      rExtrapolationMatrix,
+                                            SizeType           NumberOfIntegrationPoints,
+                                            const U&           rAtomicAddOperation)
+    {
+        auto&          r_this_geometry = rElement.GetGeometry();
+        std::vector<T> values_on_integration_points(NumberOfIntegrationPoints);
+        rElement.CalculateOnIntegrationPoints(rVariable, values_on_integration_points,
+                                              mrModelPart.GetProcessInfo());
+
+        for (IndexType iNode = 0; iNode < r_this_geometry.PointsNumber(); ++iNode) {
+            // We first initialize the source, which we need to do by getting the first value,
+            // because we don't know the size of the dynamically allocated Vector/Matrix
+            T source = rExtrapolationMatrix(iNode, 0) * values_on_integration_points[0];
+            for (IndexType i_gauss_point = 1; i_gauss_point < values_on_integration_points.size(); ++i_gauss_point) {
+                source += rExtrapolationMatrix(iNode, i_gauss_point) * values_on_integration_points[i_gauss_point];
+            }
+            source /= r_this_geometry[iNode].GetValue(mrAverageVariable);
+
+            rAtomicAddOperation(r_this_geometry[iNode].FastGetSolutionStepValue(rVariable), source);
+        }
+    }
+
+    Matrix             GetExtrapolationMatrix(const Element& rElement) const;
```

Comment:
To avoid unnecessary copies I would like the return type to be `const Matrix&` rather than `Matrix`:
```suggestion
    const Matrix&      GetExtrapolationMatrix(const Element& rElement) const;
```

Reply:
Good suggestion, didn't think when I changed the function to always return the cached matrix (and calculate it first if it doesn't exist yet). This also means the `GetCachedExtrapolationMatrixFor` now returns const Matrix&